### PR TITLE
[WIP] Support for NetStandard 1.6

### PR DESCRIPTION
**Pull request by alexvaluyskiy**

### CODE REVIEW

File: src/core/Akka/Actor/Props.cs [hunk 1]
```diff
@@ -781,7 +782,12 @@ private DynamicProps(Props copy, Func<TActor> invoker)
         protected override Props Copy()
         {
             Props initialCopy = base.Copy();
+#if CLONEABLE
             var invokerCopy = (Func<TActor>)invoker.Clone();
+#else
+            // TODO: CORECLR FIX IT
+            var invokerCopy = (Func<TActor>)invoker;
```
Comment (alexvaluyskiy) [inline src/core/Akka/Actor/Props.cs]: This is a temporary solution

File: src/core/Akka/IO/SocketChannel.cs [hunk 2]
```diff
@@ -96,6 +98,31 @@ public virtual bool IsOpen()
         /// <returns>TBD</returns>
         public bool Connect(EndPoint address)
         {
+#if CORECLR
```
Comment (alexvaluyskiy) [inline src/core/Akka/IO/SocketChannel.cs]: This is a temporary solution. Waiting for this https://github.com/akkadotnet/akka.net/pull/2405

**Comment by Aaronontheweb** [comment]
@alexvaluyskiy looks like we don't have the build system setup to pick up changes to the `netcore` branch just yet. Should I turn it on?

**Comment by alexvaluyskiy** [comment]
Yes

File: src/core/Akka.Remote/Akka.Remote.csproj [hunk 3]
```diff
@@ -1,165 +1,53 @@
...
-    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
-    <ProjectGuid>{EA4FF8FD-7C53-49C8-B9AA-02E458B3E6A7}</ProjectGuid>
-    <OutputType>Library</OutputType>
-    <AppDesignerFolder>Properties</AppDesignerFolder>
-    <RootNamespace>Akka.Remote</RootNamespace>
+    <AssemblyTitle>Akka.Remote</AssemblyTitle>
+    <Description>Remote actor support for Akka.NET</Description>
+    <Copyright>Copyright © 2013-2017 Akka.NET Team</Copyright>
+    <VersionPrefix>1.5.0</VersionPrefix>
+    <TargetFrameworks>net451;netstandard1.6</TargetFrameworks>
```
Comment (alexvaluyskiy) [inline src/core/Akka.Remote/Akka.Remote.csproj]: Will go back to `net45` when DotNetty release the new version

File: src/core/Akka.Remote/Transport/Helios/HeliosTcpTransport.cs [hunk 4]
```diff
@@ -1,328 +0,0 @@
...
-using Helios.Util;
-
-namespace Akka.Remote.Transport.Helios
-{
-    /// <summary>
-    /// INTERNAL API
-    /// </summary>
-    abstract class TcpHandlers : CommonHandlers
-    {
-        private IHandleEventListener _listener;
```
Comment (alexvaluyskiy) [inline src/core/Akka.Remote/Transport/Helios/HeliosTcpTransport.cs]: @Horusiath Do we need this file?

File: src/core/Akka.Remote/Transport/Helios/HeliosTcpTransport.cs [hunk 5]
```diff
@@ -1,328 +0,0 @@
-﻿//-----------------------------------------------------------------------
-// <copyright file="HeliosTcpTransport.cs" company="Akka.NET Project">
```
Comment (alexvaluyskiy) [inline src/core/Akka.Remote/Transport/Helios/HeliosTcpTransport.cs]: @Horusiath Do we need this file?

File: src/core/Akka.Remote/Transport/AkkaIO/AddressConverters.cs [hunk 6]
```diff
@@ -4,7 +4,7 @@
 //     Copyright (C) 2013-2016 Akka.NET project <https://github.com/akkadotnet/akka.net>
 // </copyright>
 //-----------------------------------------------------------------------
-
+#if AKKAIO
```
Comment (alexvaluyskiy) [inline src/core/Akka.Remote/Transport/AkkaIO/AddressConverters.cs]: Waiting for https://github.com/akkadotnet/akka.net/pull/2405

File: src/core/Akka.Tests.Shared.Internals/AkkaSpec.cs [hunk 7]
```diff
@@ -89,6 +90,10 @@ protected override void AfterAll()
 
         private static string GetCallerName()
         {
+#if CORECLR
```
Comment (alexvaluyskiy) [inline src/core/Akka.Tests.Shared.Internals/AkkaSpec.cs]: Have to fix that ASAO

File: src/core/Akka.Streams/Dsl/Keep.cs [hunk 8]
```diff
@@ -55,7 +56,11 @@ public static class Keep
         /// <returns>TBD</returns>
         public static NotUsed None<TLeft, TRight>(TLeft left, TRight right) => NotUsed.Instance;
 
+#if !CORECLR
         private static readonly RuntimeMethodHandle KeepRightMethodhandle = typeof(Keep).GetMethod(nameof(Right)).MethodHandle;
+#else
+        private static readonly MethodInfo KeepRightMethodInfo = typeof(Keep).GetMethod(nameof(Right));
```
Comment (alexvaluyskiy) [inline src/core/Akka.Streams/Dsl/Keep.cs]: @Silv3rcircl3 need your advice how to convert this to netcore properly

File: src/core/Akka.Streams.Tests.Performance/Akka.Streams.Tests.Performance.csproj [hunk 9]
```diff
@@ -1,126 +1,32 @@
...
-    <Compile Include="InterpreterBenchmark.cs" />
-    <Compile Include="GraphBuilderBenchmark.cs" />
-    <Compile Include="FlowSelectBenchmark.cs" />
-    <Compile Include="IO\FileSourcesBenchmark.cs" />
-    <Compile Include="JsonFramingBenchmark.cs" />
-    <Compile Include="MergeManyBenchmark.cs" />
-    <Compile Include="MaterializationBenchmark.cs" />
-    <Compile Include="Properties\AssemblyInfo.cs" />
+    <ProjectReference Include="..\Akka.Streams\Akka.Streams.csproj" />
+    <ProjectReference Include="..\Akka.Streams.Tests\Akka.Streams.Tests.csproj" />
```
Comment (alexvaluyskiy) [inline src/core/Akka.Streams.Tests.Performance/Akka.Streams.Tests.Performance.csproj]: @Silv3rcircl3 I think `Akka.Streams.Tests.Performance.csproj` should not use `Akka.Streams.Tests` and `Akka.Streams.TestKit.Tests` projects

File: src/core/Akka.Streams/Dsl/Keep.cs (hunk 8 above)
Comment (Silv3rcircl3) [inline src/core/Akka.Streams/Dsl/Keep.cs]: `IsRight` and `IsLeft` are used during materialization for [performance improvements](https://github.com/akkadotnet/akka.net/blob/4805420124154c804ff4dddcc84d9a09a4817070/src/core/Akka.Streams/Implementation/StreamLayout.cs#L730) and for [composing a materialized value with `EmptyModul` ](https://github.com/akkadotnet/akka.net/blob/4805420124154c804ff4dddcc84d9a09a4817070/src/core/Akka.Streams/Implementation/StreamLayout.cs#L955) . 

For me it looks like it _should_ work, but you should be able to verify it by debugging some tests, maybe [this one](https://github.com/akkadotnet/akka.net/blob/aa913c26cfb3f92f0a6d3ad1f24e1533cc5a9277/src/core/Akka.Streams.Tests/Dsl/GraphMatValueSpec.cs#L252) and check if one of the methods return true or the exception is thrown in the [`EmptyModule`](https://github.com/akkadotnet/akka.net/blob/4805420124154c804ff4dddcc84d9a09a4817070/src/core/Akka.Streams/Implementation/StreamLayout.cs#L955)

File: src/core/Akka.Streams.Tests.Performance/Akka.Streams.Tests.Performance.csproj (hunk 9 above)
Comment (Silv3rcircl3) [inline src/core/Akka.Streams.Tests.Performance/Akka.Streams.Tests.Performance.csproj]: You can remove the reference to `Akka.Streams.TestKit.Tests` by copying the `reference.conf` into the performance project and change [this](https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Streams.Tests.Performance/FlowSelectBenchmark.cs#L57) lines.

For the `Akka.Streams.Tests` projects it's not that easy, the [`InterpreterBenchmark`](https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Streams.Tests.Performance/InterpreterBenchmark.cs#L49) needs the [`GraphInterpreterSpecKit`](https://github.com/akkadotnet/akka.net/blob/475ff9b117e07def0c0d856a7317a32e2edc9f7c/src/core/Akka.Streams.Tests/Implementation/Fusing/GraphInterpreterSpecKit.cs) class, so the only option would be to remove this spec completely.
But I'm not sure why net452 would be a problem? NBench is currently targeting net45 and we can use any version we want for the performance project.

**Comment by alexvaluyskiy** [comment]
TODO list (create tickets)
- Fix a build of `Akka.FSharp.Tests` project (@Horusiath need your help)
- Fix MultiNode tests for net45, create MultiNode tests for netcore
- Fix Performance tests for net45, create Performance tests for netcore
- Fix `Akka.Remote.Tests` and `Akka.Remote.TestKit.Tests` projects
- Convert `Akka.DistributedData` to netstandard1.6
- Fix tests in `Akka.Persistence.Sqlite.Tests`
- Fix tests in 'Akka.Streams.Tests` (@Silv3rcircl3 need your help)

**Comment by Horusiath** [comment]
@alexvaluyskiy what about akka.io? Isn't #2405 necessary before making Akka.NET .NET core compatible?

**Comment by alexvaluyskiy** [comment]
@Horusiath AkkaIO is very important for NetCore, right now the NetCore version excludes all AkkaIO code

File: src/core/Akka/Actor/Props.cs (hunk 1 above)
Comment (tintoy) [inline src/core/Akka/Actor/Props.cs]: Not that it'll help you yet, but this is being restored in netstandard2.0 (dotnet/corefx#12251).

Comment (tintoy) [inline src/core/Akka/Actor/Props.cs]: BTW, why clone it at all - aren't delegates immutable?

**Comment by Aaronontheweb** [comment]
FYI, just released NBench 1.0 to NuGet yesterday, which now supports .NET Core and .NET Standard.

**Comment by alexvaluyskiy** [comment]
Replaced https://github.com/akkadotnet/akka.net/pull/2647

**Pull request closed without merging**